src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs

```diff
@@ -85,10 +84,17 @@ public static class ServiceCollectionDescriptorExtensions
                 throw new ArgumentNullException(nameof(descriptor));
             }
 
-            if (!collection.Any(d => d.ServiceType == descriptor.ServiceType))
+            int count = collection.Count;
+            for (int i = 0; i < count; i++)
             {
-                collection.Add(descriptor);
+                if (collection[i].ServiceType == descriptor.ServiceType)
+                {
+                    // Already added
+                    return;
+                }
```

Comment:
How many times are these methods called in a typical startup? The fact that you said it's saving ~750 closures suggests it's called to add that many services,  and this is O(N^2)?

Reply:
Yep definitely is; can get the call counts atm as Tracing in dotTrace is broken for me atm 😢

Reply:
It'd be good to understand the actual cost.  While saving 750 allocations is good,  I'd expect saving 250,000 interface dispatches to be better ;-)

Reply:
The service collection needs to store a dictionary and introduce a new interface to check if a service type already exists